src/boot/efi/pe.c

```diff
@@ -160,6 +166,46 @@ static void locate_sections(
         }
 }
 
+static UINT32 get_compat_entry_address(const struct DosFileHeader *dos, const struct PeFileHeader *pe) {
+        UINTN addr = 0, size = 0;
+        const CHAR8 *sections[] = { (CHAR8 *) ".compat", NULL };
+
+        locate_sections(
+                (struct PeSectionHeader *) ((UINT8 *) dos + section_table_offset(dos, pe)),
+                pe->FileHeader.NumberOfSections,
+                sections,
+                &addr,
+                NULL,
+                &size);
+
+        if (size == 0)
+                return 0;
+
+        typedef struct {
+                UINT8 type;
+                UINT8 size;
+                UINT16 machine_type;
+                UINT32 entry_point;
+        } _packed_ LinuxPeCompat1;
+
+        while (size >= sizeof(LinuxPeCompat1)) {
+                LinuxPeCompat1 *compat = (LinuxPeCompat1 *) ((UINT8 *) dos + addr);
+
+                if (compat->type == 0 || compat->size == 0)
+                        break;
+
+                if (compat->type == 1 &&
+                    compat->size >= sizeof(LinuxPeCompat1) &&
+                    compat->machine_type == TARGET_MACHINE_TYPE)
+                        return compat->entry_point;
+
+                addr += compat->size;
```

Comment:
overflow checks?

Reply:
The `compat->size > size` some lines above should take care of this, no?